File: python/nemo_relay/scope.py
# SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: Apache-2.0

"""Scope stack operations.

Scopes define the hierarchy that tool calls, LLM calls, and mark events attach
to. They are the main way to model agents, tasks, and nested units of work.

Example::

    import nemo_relay

    with nemo_relay.scope.scope("demo-agent", nemo_relay.ScopeType.Agent) as handle:
        nemo_relay.scope.event("checkpoint", handle=handle, data={"step": 1})
"""

from __future__ import annotations

import asyncio
from contextlib import contextmanager
from datetime import datetime
from typing import Iterator

from nemo_relay import Json
from nemo_relay._context import ensure_scope_stack as _ensure_scope_stack
from nemo_relay._native import (
    ScopeAttributes,
    ScopeHandle,
    ScopeType,
)
from nemo_relay._native import (
    event as _native_event,
)
from nemo_relay._native import (
    get_handle as _native_get_handle,
)
from nemo_relay._native import (
    pop_scope as _native_pop_scope,
)
from nemo_relay._native import (
    push_scope as _native_push_scope,
)


def get_handle() -> ScopeHandle:
    """Return the current top-of-stack ``ScopeHandle``.

    Returns:
        ScopeHandle: The scope currently at the top of the active scope stack.

    Notes:
        If the current Python context does not yet have a scope stack, one is
        created automatically before the handle lookup.
    """
    _ensure_scope_stack()
    return _native_get_handle()


def push(
    name: str,
    scope_type: ScopeType,
    *,
    handle: ScopeHandle | None = None,
    attributes: ScopeAttributes | None = None,
    data: Json | None = None,
    metadata: Json | None = None,
    input: Json | None = None,
    timestamp: datetime | None = None,
) -> ScopeHandle:
    """Push a new child scope and return its handle.

    Args:
        name: Human-readable name for the new scope.
        scope_type: Semantic scope type, such as ``ScopeType.Agent`` or
            ``ScopeType.Function``.
        handle: Optional parent scope handle. When omitted, the current
            top-of-stack scope becomes the parent.
        attributes: Optional native scope attributes attached to the emitted
            start event.
        data: Optional JSON application payload stored on the scope handle.
        metadata: Optional JSON metadata recorded on the scope start event.
        input: Optional JSON payload exported as the semantic scope input.
        timestamp: Optional timezone-aware ``datetime`` recorded as the handle
            start time and on the scope start event. When omitted, the current
            runtime time is used.

    Returns:
        ScopeHandle: Handle for the newly pushed scope.

    Notes:
        A scope stack is created automatically if the current context does not
        yet have one. ``timestamp`` must be a timezone-aware ``datetime``;
        strings and naive datetimes are rejected.

    Example::

        import nemo_relay

        with nemo_relay.scope.scope("parent", nemo_relay.ScopeType.Agent) as parent:
            handle = nemo_relay.scope.push(
                "worker",
                nemo_relay.ScopeType.Function,
                handle=parent,
                attributes=None,
                data={"step": 1},
                metadata={"source": "scope.push"},
            )
            nemo_relay.scope.pop(handle)
    """
    _ensure_scope_stack()
    return _native_push_scope(
        name,
        scope_type,
        handle=handle,
        attributes=attributes,
        data=data,
        metadata=metadata,
        input=input,
        timestamp=timestamp,
    )


def pop(
    handle: ScopeHandle, *, output: Json | None = None, metadata: Json | None = None, timestamp: datetime | None = None
) -> None:
    """Pop a scope previously returned by ``push()`` or ``scope()``.

    Args:
        handle: Scope handle to close.
        output: Optional JSON payload exported as the semantic scope output.
        metadata: Optional JSON metadata to append to the metadata set when the scope was created.
        timestamp: Optional timezone-aware ``datetime`` recorded on the scope
            end event. When omitted, the runtime default end timestamp is used.

    Returns:
        None: This function returns after the scope is closed successfully.

    Notes:
        The handle must correspond to an active scope in the current scope
        stack. Popping a scope also removes any scope-local registrations owned
        by that scope. ``timestamp`` must be a timezone-aware ``datetime``;
        strings and naive datetimes are rejected.
    """
    _ensure_scope_stack()
    _native_pop_scope(handle, output=output, metadata=metadata, timestamp=timestamp)


def event(
    name: str,
    *,
    handle: ScopeHandle | None = None,
    data: Json | None = None,
    metadata: Json | None = None,
    timestamp: datetime | None = None,
) -> None:
    """Emit a ``Mark`` event under the current or provided scope.

    Args:
        name: Event name to emit.
        handle: Optional scope handle that should own the event. When omitted,
            the current top-of-stack scope is used.
        data: Optional JSON payload attached to the event.
        metadata: Optional JSON metadata attached to the event.
        timestamp: Optional timezone-aware ``datetime`` recorded on the mark
            event. When omitted, the current runtime time is used.

    Returns:
        None: This function returns after the event has been queued for
        sanitization and publication.

    Notes:
        A scope stack is created automatically when needed before the event is
        queued through the native runtime. ``timestamp`` must be a
        timezone-aware ``datetime``; strings and naive datetimes are rejected.
    """
    _ensure_scope_stack()
    _native_event(name, handle=handle, data=data, metadata=metadata, timestamp=timestamp)


@contextmanager
def scope(
    name: str,
    scope_type: ScopeType,
    *,
    handle: ScopeHandle | None = None,
    attributes: ScopeAttributes | None = None,
    data: Json | None = None,
    metadata: Json | None = None,
    input: Json | None = None,
    timestamp: datetime | None = None,
    end_timestamp: datetime | None = None,
) -> Iterator[ScopeHandle]:
    """Create a scope for the duration of a ``with`` block.

    OTEL status codes will be automatically recorded in the scope's metadata.

    Args:
        name: Human-readable name for the new scope.
        scope_type: Semantic scope type, such as ``ScopeType.Agent`` or
            ``ScopeType.Function``.
        handle: Optional parent scope handle. When omitted, the current
            top-of-stack scope becomes the parent.
        attributes: Optional native scope attributes attached to the emitted
            start event.
        data: Optional JSON application payload stored on the scope handle.
        metadata: Optional JSON metadata recorded on the scope start event.
        input: Optional JSON payload exported as the semantic scope input.
        timestamp: Optional timezone-aware ``datetime`` recorded as the handle
            start time and on the scope start event.
        end_timestamp: Optional timezone-aware ``datetime`` recorded on the
            scope end event.

    Yields:
        ScopeHandle: Handle for the scope that remains active inside the
        ``with`` block.

    Notes:
        The scope is always popped when the ``with`` block exits, even if the
        body raises an exception. Timestamp arguments must be timezone-aware
        ``datetime`` objects; strings and naive datetimes are rejected.

    Example::

        import nemo_relay

        with nemo_relay.scope.scope(
            "demo",
            nemo_relay.ScopeType.Agent,
            handle=None,
            attributes=None,
            data={"stage": "start"},
            metadata={"owner": "docs"},
        ) as handle:
            nemo_relay.scope.event("inside", handle=handle, data={"ok": True}, metadata={"step": 1})
    """
    _ensure_scope_stack()
    pushed_handle = None
    status_code = "UNSET"
    status_message = None
    try:
        pushed_handle = _native_push_scope(
            name,
            scope_type,
            handle=handle,
            attributes=attributes,
            data=data,
            metadata=metadata,
            input=input,
            timestamp=timestamp,
        )
        yield pushed_handle
        status_code = "OK"
    except asyncio.CancelledError as e:
        status_code = "ERROR"
        status_message = str(e) or "cancelled"
        raise
    except Exception as e:
        status_code = "ERROR"
        status_message = str(e)
        raise
    finally:
        if pushed_handle is not None:
            metadata = {"otel.status_code": status_code}
            if status_message is not None:
                metadata["otel.status_description"] = status_message
            pop(pushed_handle, metadata=metadata, timestamp=end_timestamp)


__all__ = ["event", "get_handle", "pop", "push", "scope"]

File: crates/core/src/api/scope.rs
// SPDX-FileCopyrightText: Copyright (c) 2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::api::event::{
    BaseEvent, CategoryProfile, DataSchema, EventCategory, LOG_SEVERITY_METADATA_KEY, LogSeverity,
    METRIC_DATA_SCHEMA_NAME, METRIC_DATA_SCHEMA_VERSION, MarkEvent, MetricEnvelope,
    MetricMeasurement,
};
use crate::api::runtime::global_context;
use crate::api::runtime::scope_stack::snapshot_scope_stack;
use crate::api::runtime::subscriber_dispatcher::{self, SubscriberDelivery};
use crate::api::runtime::{
    current_scope_stack, task_scope_push, task_scope_remove, task_scope_top,
};
use crate::api::shared::{
    ensure_runtime_owner, resolve_parent_uuid, snapshot_event_sanitizers,
    snapshot_event_subscribers,
};
use crate::error::{FlowError, Result};
use crate::json::Json;
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use typed_builder::TypedBuilder;
use uuid::Uuid;

pub use nemo_relay_types::api::scope::{HandleAttributes, ScopeAttributes, ScopeType};

/// Canonical mark-event name used to indicate agent context compaction.
pub const COMPACTION_EVENT_NAME: &str = "compaction";

/// Runtime-owned handle identifying an active or completed scope.
#[derive(Debug, Clone, Serialize, Deserialize, TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct ScopeHandle {
    /// Unique scope identifier.
    #[builder(default = Uuid::now_v7())]
    pub uuid: Uuid,
    /// Timestamp captured when the scope handle was created.
    #[builder(default = Utc::now())]
    pub started_at: DateTime<Utc>,
    /// Semantic category of the scope.
    pub scope_type: ScopeType,
    /// Human-readable scope name.
    #[builder(setter(into))]
    pub name: String,
    /// Optional application payload stored on the handle.
    #[builder(default)]
    pub data: Option<Json>,
    /// Optional metadata attached to the scope.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Scope behavior flags.
    #[builder(default = ScopeAttributes::empty())]
    pub attributes: ScopeAttributes,
    /// UUID of the parent scope, if any.
    #[builder(default)]
    pub parent_uuid: Option<Uuid>,
}

fn scope_stack_lock_error(error: impl std::fmt::Display, operation: &'static str) -> FlowError {
    log::error!(
        target: "nemo_relay.runtime",
        event = "scope_stack_unavailable",
        operation = operation;
        "Scope operation failed because the scope stack lock is poisoned: {error}"
    );
    FlowError::Internal(error.to_string())
}

/// Builder parameters for [`push_scope`].
#[derive(TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct PushScopeParams<'a> {
    /// Human-readable scope name recorded on emitted lifecycle events.
    pub name: &'a str,
    /// Semantic category for the new scope.
    pub scope_type: ScopeType,
    /// Optional explicit parent scope.
    #[builder(default)]
    pub parent: Option<&'a ScopeHandle>,
    /// Scope attribute bitflags applied to the new scope.
    #[builder(default = ScopeAttributes::empty())]
    pub attributes: ScopeAttributes,
    /// Optional application payload stored on the scope handle.
    #[builder(default)]
    pub data: Option<Json>,
    /// Optional JSON metadata recorded on the emitted start event.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Optional JSON payload exported as the scope start event data.
    #[builder(default)]
    pub input: Option<Json>,
    /// Optional timestamp recorded on the emitted start event.
    #[builder(default)]
    pub timestamp: Option<DateTime<Utc>>,
}

/// Builder parameters for [`NemoRelayContextState::create_scope_handle`].
#[derive(Debug, Clone, TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct CreateScopeHandleParams<'a> {
    /// Human-readable scope name.
    pub name: &'a str,
    /// Optional parent scope UUID.
    #[builder(default)]
    pub parent_uuid: Option<Uuid>,
    /// Semantic category of the scope.
    pub scope_type: ScopeType,
    /// Scope attribute bitflags.
    #[builder(default = ScopeAttributes::empty())]
    pub attributes: ScopeAttributes,
    /// Optional application payload stored on the handle.
    #[builder(default)]
    pub data: Option<Json>,
    /// Optional metadata stored on the handle.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Optional timestamp captured as the handle start time and reused by the
    /// emitted start event. When omitted, the current UTC time is used.
    #[builder(default)]
    pub timestamp: Option<DateTime<Utc>>,
}

/// Builder parameters for [`NemoRelayContextState::build_scope_end_event`].
#[derive(Debug, Clone, TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct EndScopeHandleParams<'a> {
    /// Scope handle to serialize into the emitted end event.
    pub handle: &'a ScopeHandle,
    /// Optional JSON payload exported as the semantic scope output.
    #[builder(default)]
    pub data: Option<Json>,
    /// Optional metadata to be appended to the metadata set when the scope was created.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Optional timestamp recorded on the emitted end event. When omitted, the
    /// runtime records the current UTC time, or one microsecond after the
    /// handle start time if the current time is not later.
    #[builder(default)]
    pub timestamp: Option<DateTime<Utc>>,
}

/// Builder parameters for [`pop_scope`].
#[derive(TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct PopScopeParams<'a> {
    /// UUID of the scope that should be popped.
    pub handle_uuid: &'a Uuid,
    /// Optional JSON payload exported as the semantic scope output.
    #[builder(default)]
    pub output: Option<Json>,
    /// Optional JSON payload metadata to be appended to the metadata set when the scope was created.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Optional timestamp recorded on the emitted end event. When omitted, the
    /// runtime records the current UTC time, or one microsecond after the
    /// handle start time if the current time is not later.
    #[builder(default)]
    pub timestamp: Option<DateTime<Utc>>,
}

/// Builder parameters for [`event`].
#[derive(TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct EmitMarkEventParams<'a> {
    /// Event name to emit.
    pub name: &'a str,
    /// Optional explicit parent scope.
    #[builder(default)]
    pub parent: Option<&'a ScopeHandle>,
    /// Optional JSON payload recorded as the mark data.
    #[builder(default)]
    pub data: Option<Json>,
    /// Optional schema identifier for the mark data.
    #[builder(default)]
    pub data_schema: Option<DataSchema>,
    /// Optional JSON metadata recorded on the emitted event.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Optional typed log severity stored authoritatively in mark metadata.
    #[builder(default)]
    pub severity: Option<LogSeverity>,
    /// Optional semantic category for the mark.
    #[builder(default)]
    pub category: Option<EventCategory>,
    /// Optional category-specific mark profile.
    #[builder(default)]
    pub category_profile: Option<CategoryProfile>,
    /// Optional timestamp recorded on the emitted mark event. When omitted, the
    /// current UTC time is used.
    #[builder(default)]
    pub timestamp: Option<DateTime<Utc>>,
}

/// Builder parameters for [`metric`].
#[derive(TypedBuilder)]
#[builder(field_defaults(setter(strip_option(ignore_invalid, fallback_suffix = "_opt"))))]
pub struct EmitMetricEventParams<'a> {
    /// Mark name to emit for the metric recording operations.
    pub name: &'a str,
    /// Metric measurements recorded atomically by downstream metric exporters.
    pub measurements: Vec<MetricMeasurement>,
    /// Optional explicit parent scope.
    #[builder(default)]
    pub parent: Option<&'a ScopeHandle>,
    /// Optional JSON metadata recorded on the emitted event.
    #[builder(default)]
    pub metadata: Option<Json>,
    /// Optional timestamp recorded on the emitted metric mark.
    #[builder(default)]
    pub timestamp: Option<DateTime<Utc>>,
}

pub(crate) fn metadata_with_log_severity(
    metadata: Option<Json>,
    severity: Option<LogSeverity>,
) -> Result<Option<Json>> {
    let Some(severity) = severity else {
        return Ok(metadata);
    };
    let mut metadata = metadata.unwrap_or_else(|| Json::Object(serde_json::Map::new()));
    let object = metadata.as_object_mut().ok_or_else(|| {
        FlowError::InvalidArgument(
            "mark metadata must be a JSON object when severity is provided".into(),
        )
    })?;
    object.insert(
        LOG_SEVERITY_METADATA_KEY.into(),
        Json::String(severity.as_str().into()),
    );
    Ok(Some(metadata))
}

/// Return the current scope at the top of the active stack.
///
/// This reads the task-local or thread-local scope stack without mutating it
/// and returns a clone of the current top-most [`ScopeHandle`].
///
/// # Returns
/// A [`Result`] containing the current [`ScopeHandle`] when the runtime owner
/// check succeeds.
///
/// # Errors
/// Returns an error when the current binding has not initialized the shared
/// runtime ownership correctly.
pub fn get_handle() -> Result<ScopeHandle> {
    ensure_runtime_owner()?;
    Ok(task_scope_top())
}

/// Push a new scope onto the active scope stack.
///
/// This creates a new [`ScopeHandle`], emits a scope-start event to global and
/// scope-local subscribers, and makes the new scope the current top of stack.
///
/// # Parameters
/// - `name`: Human-readable scope name recorded on emitted lifecycle events.
/// - `scope_type`: Semantic category for the new scope.
/// - `parent`: Optional explicit parent scope. When `None`, the current top of
///   stack is used as the parent.
/// - `attributes`: Bitflags that modify scope behavior and observability.
/// - `data`: Optional application payload stored on the returned handle.
/// - `metadata`: Optional JSON metadata recorded on the emitted start event.
/// - `input`: Optional JSON payload exported as the Agent Trajectory
///   Observability Format (ATOF) data payload.
/// - `timestamp`: Optional timestamp recorded as the handle start time and on
///   the emitted start event. When `None`, the current UTC time is used.
///
/// # Returns
/// A [`Result`] containing the newly created [`ScopeHandle`].
///
/// # Errors
/// Returns an error when the runtime owner check fails or when internal state
/// cannot be read safely.
///
/// # Notes
/// The start event is queued with subscriber and sanitizer snapshots captured
/// while the new scope is active.
pub fn push_scope(params: PushScopeParams<'_>) -> Result<ScopeHandle> {
    ensure_runtime_owner()?;
    let parent_uuid = resolve_parent_uuid(params.parent);
    let (handle, event, subscribers, emission_scope_stack) = {
        let scope_stack = current_scope_stack();
        let scope_guard = scope_stack
            .read()
            .map_err(|error| scope_stack_lock_error(error, "push"))?;
        let scope_subscribers = scope_guard.collect_scope_local_subscribers();
        let subscribers = snapshot_event_subscribers(scope_subscribers)?;
        let context = global_context();
        let state = context
            .read()
            .map_err(|error| FlowError::Internal(error.to_string()))?;
        let handle_params = CreateScopeHandleParams::builder()
            .name(params.name)
            .parent_uuid_opt(parent_uuid)
            .scope_type(params.scope_type)
            .attributes(params.attributes)
            .data_opt(params.data)
            .metadata_opt(params.metadata)
            .timestamp_opt(params.timestamp)
            .build();
        let handle = state.create_scope_handle(handle_params);
        let event = state.build_scope_start_event(&handle, params.input);
        (handle, event, subscribers, scope_stack.clone())
    };
    task_scope_push(handle.clone());
    let sanitizers = snapshot_event_sanitizers(&event, &emission_scope_stack).unwrap_or_default();
    let _ = subscriber_dispatcher::dispatch_sanitized_event(
        event,
        sanitizers,
        &subscribers,
        emission_scope_stack,
    );
    Ok(handle)
}

/// Pop the current scope from the active scope stack.
///
/// This emits a scope-end event for the target scope and removes any
/// scope-local registrations owned by that scope.
///
/// # Parameters
/// - `handle_uuid`: UUID of the scope that should be popped.
/// - `output`: Optional JSON payload exported as the semantic scope output.
/// - `timestamp`: Optional timestamp recorded on the emitted end event. When
///   `None`, the runtime uses the current UTC time, or one microsecond after
///   the handle start time if the current time is not later.
///
/// # Returns
/// A [`Result`] that is `Ok(())` when the scope was popped successfully.
///
/// # Errors
/// Returns [`FlowError::InvalidArgument`] when the target scope exists but is
/// not the current top of stack, and [`FlowError::NotFound`] when the UUID is
/// unknown to the active stack.
///
/// # Notes
/// The implicit root scope cannot be removed.
///
/// Scope-end emission snapshots the visible scope-local sanitizers before
/// removing the scope. Publication is then queued after removal using that
/// snapshot, so cleanup does not change the middleware applied to the emitted
/// event.
pub fn pop_scope(params: PopScopeParams<'_>) -> Result<()> {
    pop_scope_inner(params, false).map(|_| ())
}

/// Pop the current scope and return a receipt for its scope-end subscriber delivery.
///
/// The receipt covers sanitizer and subscriber processing for the scope-end event.
/// It does not wait for unrelated events queued after that event.
#[doc(hidden)]
pub fn pop_scope_with_subscriber_delivery(
    params: PopScopeParams<'_>,
) -> Result<SubscriberDelivery> {
    pop_scope_inner(params, true)?.ok_or_else(|| {
        FlowError::Internal("tracked scope pop did not create a subscriber delivery receipt".into())
    })
}

fn pop_scope_inner(
    params: PopScopeParams<'_>,
    track_delivery: bool,
) -> Result<Option<SubscriberDelivery>> {
    ensure_runtime_owner()?;
    let scope_stack = current_scope_stack();
    let (scope, event, subscribers, emission_scope_stack) = {
        let scope_guard = scope_stack
            .read()
            .map_err(|error| scope_stack_lock_error(error, "pop"))?;
        let top = scope_guard.top();
        if top.uuid != *params.handle_uuid {
            if scope_guard.find(params.handle_uuid).is_some() {
                return Err(FlowError::InvalidArgument(
                    "scope handle is not at the top of the stack".into(),
                ));
            }
            return Err(FlowError::NotFound("scope handle not found".into()));
        }
        let scope_subscribers = scope_guard.collect_scope_local_subscribers();
        let subscribers = snapshot_event_subscribers(scope_subscribers)?;
        let scope = top.clone();
        let context = global_context();
        let state = context
            .read()
            .map_err(|error| FlowError::Internal(error.to_string()))?;
        let event = state.build_scope_end_event(
            EndScopeHandleParams::builder()
                .handle(&scope)
                .data_opt(params.output)
                .timestamp_opt(params.timestamp)
                .metadata_opt(params.metadata)
                .build(),
        );
        (scope, event, subscribers, scope_stack.clone())
    };
    // Capture the scope-local chain before removing its owner. The event is
    // published later, but scope cleanup must not change the middleware that
    // was visible when the end event was emitted.
    let sanitizers = snapshot_event_sanitizers(&event, &emission_scope_stack).unwrap_or_default();
    let publication_scope_stack = snapshot_scope_stack(&emission_scope_stack)?;
    let removed = task_scope_remove(params.handle_uuid)?;
    debug_assert_eq!(removed.uuid, scope.uuid);
    if track_delivery {
        subscriber_dispatcher::dispatch_sanitized_event_with_delivery(
            event,
            sanitizers,
            &subscribers,
            publication_scope_stack,
        )
        .map(Some)
    } else {
        let _ = subscriber_dispatcher::dispatch_sanitized_event(
            event,
            sanitizers,
            &subscribers,
            publication_scope_stack,
        );
        Ok(None)
    }
}

/// Emit a standalone mark event under the current or provided scope.
///
/// This creates a point-in-time lifecycle event without pushing or popping a
/// new scope.
///
/// # Parameters
/// - `name`: Event name to emit.
/// - `parent`: Optional explicit parent scope. When `None`, the current top of
///   stack is used.
/// - `data`: Optional JSON payload recorded on the emitted event.
/// - `data_schema`: Optional name and version identifying the data payload.
/// - `metadata`: Optional JSON metadata recorded on the emitted event.
/// - `severity`: Optional typed severity stored in reserved mark metadata.
/// - `category`: Optional semantic mark category.
/// - `category_profile`: Optional category-specific profile.
/// - `timestamp`: Optional timestamp recorded on the emitted mark event. When
///   `None`, the current UTC time is used.
///
/// # Returns
/// A [`Result`] that is `Ok(())` after the event has been queued for
/// sanitization and publication.
///
/// # Errors
/// Returns an error when the runtime owner check fails or when internal state
/// cannot be read safely. Returns [`FlowError::InvalidArgument`] when a typed
/// severity is provided with non-object metadata.
///
/// # Notes
/// The mark event is queued with subscriber and sanitizer snapshots captured
/// from the active scope stack.
pub fn event(params: EmitMarkEventParams<'_>) -> Result<()> {
    ensure_runtime_owner()?;
    let parent_uuid = resolve_parent_uuid(params.parent);
    let metadata = metadata_with_log_severity(params.metadata, params.severity)?;
    let scope_stack = current_scope_stack();
    let (event, subscribers, emission_scope_stack) = {
        let subscribers = if params.name == COMPACTION_EVENT_NAME {
            let mut scope_guard = scope_stack
                .write()
                .map_err(|error| scope_stack_lock_error(error, "mark"))?;
            let subscribers =
                snapshot_event_subscribers(scope_guard.collect_scope_local_subscribers())?;
            scope_guard.mark_agent_fresh(parent_uuid);
            subscribers
        } else {
            let scope_guard = scope_stack
                .read()
                .map_err(|error| scope_stack_lock_error(error, "mark"))?;
            snapshot_event_subscribers(scope_guard.collect_scope_local_subscribers())?
        };
        let context = global_context();
        let state = context
            .read()
            .map_err(|error| FlowError::Internal(error.to_string()))?;
        let event = state.create_event(MarkEvent::new(
            BaseEvent::builder()
                .name(params.name)
                .parent_uuid_opt(parent_uuid)
                .timestamp(params.timestamp.unwrap_or_else(Utc::now))
                .data_opt(params.data)
                .data_schema_opt(params.data_schema)
                .metadata_opt(metadata)
                .build(),
            params.category,
            params.category_profile,
        ));
        (event, subscribers, scope_stack.clone())
    };
    let sanitizers = snapshot_event_sanitizers(&event, &emission_scope_stack).unwrap_or_default();
    let _ = subscriber_dispatcher::dispatch_sanitized_event(
        event,
        sanitizers,
        &subscribers,
        emission_scope_stack,
    );
    Ok(())
}

/// Emit a Relay metric mark under the current or provided scope.
///
/// The measurements are validated as one atomic envelope before the mark is
/// queued. The emitted mark uses the Relay-owned metric data schema so metric
/// exporters can route it without relying on mutable metadata.
///
/// # Errors
/// Returns [`FlowError::InvalidArgument`] when the measurement envelope is
/// invalid, or any error returned while emitting the underlying mark.
pub fn metric(params: EmitMetricEventParams<'_>) -> Result<()> {
    ensure_runtime_owner()?;
    let envelope = MetricEnvelope {
        measurements: params.measurements,
    };
    envelope
        .validate()
        .map_err(|error| FlowError::InvalidArgument(error.to_string()))?;
    let data = serde_json::to_value(envelope).map_err(|error| {
        FlowError::InvalidArgument(format!("metric envelope could not be serialized: {error}"))
    })?;
    event(
        EmitMarkEventParams::builder()
            .name(params.name)
            .parent_opt(params.parent)
            .data(data)
            .data_schema(
                DataSchema::builder()
                    .name(METRIC_DATA_SCHEMA_NAME)
                    .version(METRIC_DATA_SCHEMA_VERSION)
                    .build(),
            )
            .metadata_opt(params.metadata)
            .timestamp_opt(params.timestamp)
            .build(),
    )
}

#[cfg(test)]
#[path = "../../tests/unit/scope_api_tests.rs"]
mod tests;
